linkedlist/singly/go/singlylinkedlist.go

```diff
@@ -0,0 +1,187 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+// Node is an element in the linked list
+type Node struct {
+	element interface{}
+	next    *Node
+}
+
+// NewNode returns a new node
+func NewNode(element interface{}) *Node {
+	return &Node{
+		element: element,
+		next:    nil,
+	}
+}
+
+// SinglyLinkedList is a unidirectional linked list
+type SinglyLinkedList struct {
+	head *Node
+	tail *Node
+	size int
+}
+
+// NewSinglyLinkedList returns a new singly-linked list
+func NewSinglyLinkedList() *SinglyLinkedList {
+	return &SinglyLinkedList{}
+}
+
+// AddFirst adds an element to the front of the list so that it becomes the new head node
+func (l *SinglyLinkedList) AddFirst(element interface{}) {
+	newHead := NewNode(element)
+	newHead.next = l.head
+	l.head = newHead
+
+	if l.tail == nil {
+		l.tail = l.head
+	}
+
+	l.size++
+}
+
+// AddLast adds an element to the back of the list so that it becomes the new tail node
+func (l *SinglyLinkedList) AddLast(element interface{}) {
+	node := l.head
+	for node.next != nil {
+		node = node.next
+	}
+
+	node.next = NewNode(element)
+	l.tail = node.next
+
+	l.size++
+}
+
+// AddAtPosition adds an element to a specified position in the list
+func (l *SinglyLinkedList) AddAtPosition(n int, element interface{}) error {
+	if n < 0 || n > l.size {
```

Comment:
this should be `n < 1`, we are using the 1-indexed approach because it's easier to understand.

Also updated the Java impl. See: https://github.com/oyekanmiayo/data-structures-all-langs/blob/main/linkedlist/singly/java/SinglyLinkedList.java#L99

Reply:
Sure. I think I got a bit confused by the Java implementation. Wasn't sure if it was 0 or 1 based. Will fix.